openmdao/components/submodel_comp.py

```diff
@@ -0,0 +1,447 @@
+"""Define the SubmodelComp class for evaluating OpenMDAO systems within components."""
+
+from openmdao.core.explicitcomponent import ExplicitComponent
+from openmdao.utils.general_utils import find_matches
+from openmdao.core.constants import _SetupStatus, INF_BOUND
+
+
+class SubmodelComp(ExplicitComponent):
+    """
+    System level container for systems.
+
+    Parameters
+    ----------
+    model : <System>
+        The system-level <System>.
+    problem : object
+        Instantiated problem to use for the model.
+    inputs : list of str or tuple or None
+        List of provided input names in str or tuple form. If an element is a str,
+        then it should be the promoted name in its group. If it is a tuple,
+        then the first element should be the group's promoted name, and the
+        second element should be the var name you wish to refer to it within the subproblem
+        [e.g. (path.to.var, desired_name)].
+    outputs : list of str or tuple or None
+        List of provided output names in str or tuple form. If an element is a str,
+        then it should be the promoted name in its group. If it is a tuple,
+        then the first element should be the group's promoted name, and the
+        second element should be the var name you wish to refer to it within the subproblem
+        [e.g. (path.to.var, desired_name)].
+    reports : bool
+        Determines if reports should be include in subproblem. Default is False because
+        submodelcomp runs faster without reports.
+    **kwargs : named args
+        All remaining named args that become options for `SubmodelComp`.
+
+    Attributes
+    ----------
+    _subprob : object
+        Instantiated problem used to run the model.
+    submodel_inputs : list of tuple
+        List of inputs requested by user to be used as inputs in the
+        subproblem's system.
+    submodel_outputs : list of tuple
+        List of outputs requested by user to be used as outputs in the
+        subproblem's system.
+    is_set_up : bool
+        Flag to determne if subproblem is set up. Used for add_input/add_output to
+        determine how to add the io.
+    """
+
+    def __init__(self, problem, inputs=None, outputs=None, reports=False, **kwargs):
+        """
+        Initialize all attributes.
+        """
+        super().__init__(**kwargs)
+
+        problem._reports = [] if not reports else problem._reports
+        # promote ever var in subsystems automatically
+        for subsys in problem.model._static_subsystems_allprocs.keys():
+            problem.model.promotes(subsys, any=['*'])
```

Comment:
It's not clear to me why we're promoting everything here.  It could also cause errors due to collisions between outputs that have the same name.

Reply:
My original thinking was that we want everything to be promoted so that it's easier to refer to the submodel's variables inside. But you're right, that would cause collisions with outputs that have the same name and if the user wants to promote variables, they can do that using `problem.model.promotes` anyway. I'll go ahead and change that.

Reply:
got rid of automatic promoting